### ISSUE TITLE
Unicode fix - beta testers needed

### ISSUE THREAD

**Issue by GammaC0de** [pyload/pyload#3487]
Even though python 2 now reaches end of life,
pyLoad v0.5 is not going to be finished in the near future :cry: 

Because of this I finnaly took the effort to try to fix the Unicode issues (umlaut filenames / 'ascii' codec can't decode).
Also fixed in this version Unicode parsing of `Content-Disposition` HTTP header.

So in order to test my work,
I need you to either:
- Post here URLs that contains umlauts filenames or even URLs with foreign languages filenames like Chinese.
- Try it yourself and report issues:
 Download the [python2-unicode-fix](https://github.com/GammaC0de/pyload/archive/python2-unicode-fix.zip) branch and update your installation using the instructions in https://github.com/pyload/pyload/issues/3283.

**Comment by keqas** [comment]
How can I substitute the files in a Docker container? (Synology DSM, linuxserver/pyload)

**Comment by GammaC0de** [comment]
@keqas I am not familiar with Docker,
Can you try installing this branch using entwere?

**Comment by haroldmrt** [comment]
@keqas you can follow the [Building locally](https://github.com/linuxserver/docker-pyload#building-locally) section of linuxserver readme. You will have to change the archive URL:

```
sed -i "s/https:\/\/github.com\/pyload\/pyload\/archive\/\${PYLOAD_VERSION}.tar.gz/https:\/\/github.com\/GammaC0de\/pyload\/archive\/python2-unicode-fix.tar.gz/"
```
After that I think you still have to delete the `config/userplugins` directory.

@GammaC0de it seems that I cannot add an account while running this branch: 
I get this error (even after deleting config/accounts.conf and adding the account):

```
Error importing RealdebridCom: cannot import name fs_encode
```
[error_log_account.txt](https://github.com/pyload/pyload/files/3983023/error_log_account.txt)


I briefly looked at your MR and I have this inside my docker container, if you need the info:
```
>>> sys.getfilesystemencoding()
'UTF-8'
```

**Comment by haroldmrt** [comment]
Here is the image I built manually if needed (linuxserver/pyload): https://drive.google.com/file/d/1GpR0S8hqEF8HKMTODtMv1rDjpdtCLRmU/view?usp=sharing

I created a dummy torrent file triggering encoding bugs on stable (contains a file name with unicode character 'é'). It stills does not work on python2-unicode-fix. Logs are [here](https://github.com/pyload/pyload/files/3987786/pyload_unicode_test.zip).

**Comment by keqas** [comment]
@haroldmrt
Thank you for providing us with a customized docker image! However, I succeeded in building my own image only about half an hour before your post :)

@GammaC0de
I can confirm that downloading files from uploaded.net, rapidgator.net and ddl.to is now working fine in combination with premium.to: premium mode is working; file names are pulled correctly, hence auto extraction is working as well. -> Thank you very much!
But it seems that there is a former bug coming up again:
When adding links via ClickNLoad they are put into the collector although the respective CnL  parameter is set to "queue".

**Comment by GammaC0de** [comment]
@haroldmrt @keqas Thanks for for your input and effort ;)

I am sure the is more that has to be done before we can merge this into stable,
This is why #3487 was opened. :)

>When adding links via ClickNLoad they are put into the collector although the respective CnL parameter is set to "queue".

@keqas Can you provide more info like sample URL / debug log?

**Comment by keqas** [comment]
Try this URL: https://ddl-warez.to/download/513059/Kodi.v18.0.Leia.Portable.Multilanguage-P2P/
-> Rapidgator -> Download via Click'n'load
[My Debug Logs & Screenshots](https://github.com/pyload/pyload/files/3988704/CnL.Debug.zip)
(Collector = Linksammler, Queue = Warteschlange ;)

**Comment by GammaC0de** [comment]
@haroldmrt @keqas
I have Added more fixes to the branch, please update your images and clear the userplugins dircetory.

@haroldmrt
Torrent should work fine now, please test and confirm.

@keqas
 I have tried ClickNLoad and it works fine for me..

The way ClickNLoad works in pyLoad is that it first added to the collector and only after that is is moved to the queue if it configured to go to the queue.
So probably something happend after the package added to the collector and before it was moved to the queue, 
maybe you have stopped pyLoad then?

Can you retry again please?

**Comment by haroldmrt** [comment]
Testing with same torrent file, this time the issue is with urllib.
Here is the stack trace:
```
TRACEBACK:
 Traceback (most recent call last):
  File "/app/pyload/module/PluginThread.py", line 363, in run
    self.active.plugin.preprocessing(self)
  File "/app/pyload/module/plugins/internal/Base.py", line 294, in preprocessing
    return self._process(*args, **kwargs)
  File "/app/pyload/module/plugins/internal/Base.py", line 284, in _process
    self.process(self.pyfile)
  File "/app/pyload/module/plugins/internal/Container.py", line 34, in process
    self.decrypt(pyfile)
  File "/app/pyload/module/plugins/container/TORRENT.py", line 44, in decrypt
    self.packages.append((pack_name, ["file://%s" % urllib.pathname2url(torrent_filename)], pack_name))
  File "/usr/lib/python2.7/urllib.py", line 61, in pathname2url
    return quote(pathname)
  File "/usr/lib/python2.7/urllib.py", line 1301, in quote
    return ''.join(map(quoter, s))
KeyError: u'\xe9'
```

**Comment by GammaC0de** [comment]
@haroldmrt Can you provide sample URL?

**Comment by haroldmrt** [comment]
I provided the torrent file above in a zip:
https://github.com/pyload/pyload/files/3987786/pyload_unicode_test.zip
I uploaded it as a container.
It is a torrent file containing only a filename with some Unicode
characters.

**Comment by GammaC0de** [comment]
@haroldmrt Branch updated, please retest

**Comment by haroldmrt** [comment]
Here are my test results:

**First test:** downloading 'dict.torrent' (containing Dictionnaire français-turc.djvu): it works ! However I still have some errors in logs, you can have a look at it and see if it is meaningful.

**Second test:** downloading 'Dictionnaire français-turc.torrent' (renamed first test torrent file). It does not work because of the torrent file name, see logs.

**Third test:** downloading same torrent with a magnet link: it works.

All torrents and log files are here: [pyload_unicode_torrents.zip](https://github.com/pyload/pyload/files/3998550/pyload_unicode_torrents.zip)

**Comment by GammaC0de** [comment]
Oopsi  :flushed: 

Please test now.

**Comment by haroldmrt** [comment]
Second test is now fully working. This issue looks fixed for my use cases :)

**Comment by keqas** [comment]
There was a misconfiguration concerning the port forwarding rules for CnL in my setup related with docker. After renewing the port mapping and forwarding rules CnL now works fine for me!
So I finally can confirm that this version fixed my issue #3457. Thank you!

**Comment by AlexPi86** [comment]
**GammaC0de**
Thank you for your fix.
I've downloaded pyload-python2-unicode-fix.zip
and installed it using instuctions in #3283
But when I start pyload I see an error:

/opt/share # /opt/etc/init.d/S51pyload start
starting pyLoad
/opt/etc/init.d/S51pyload: line 8: pyload: not found

Then I decided to rollback my initial folder of pyload (I've backed up it to /opt/share/pyload_entware_package)
And after doing that I have new errors:

starting pyLoad
Traceback (most recent call last):
  File "/opt/bin/pyload", line 42, in <module>
    from module import InitHomeDir
  File "/opt/share/pyload/module/InitHomeDir.py", line 28, in <module>
    builtin.owd = path.abspath("") #original working directory
  File "/opt/lib/python2.7/posixpath.py", line 371, in abspath
OSError: [Errno 2] No such file or directory

I've made 
opkg remove pyload 
opkg install pyload

Now it works.

I dont understand why replacing /opt/share/pyload with files from pyload-python2-unicode-fix.zip
breakes pyload start **(/opt/etc/init.d/S51pyload: line 8: pyload: not found)** 
Line 8 is: pyload --daemon --configdir=/opt/var/pyload
pyload  is symlink in /opt/bin to the file /opt/share/pyload/pyLoadCore.py
which is OK

**Comment by GammaC0de** [comment]
@AlexPi86 Cany you please open a new issue and I will try to help you there, ok?

**Comment by AlexPi86** [comment]
**@GammaC0de** 
Thank you!
#3501

**Comment by GammaC0de** [comment]
This is now v0.4.20 (#3520)
